lib/Transforms/IPO/StructAnalysisCloseProximity.cpp

```diff
@@ -0,0 +1,671 @@
+          auto C = 0;
+          if (auto ExCnt = StructInfo->getExecutionCount(&I))
+            C = ExCnt.getValue();
+          auto D = BBI->RemainBytes;
+          FRG->connectNodes(BBI->LastNode, NewNode, C, D);
+          BBI->RemainBytes = 0;
+          DEBUG_WITH_TYPE(DEBUG_TYPE_FRG, dbgs() << "Connect new node with previous node in the BB: Field [" << BBI->LastNode->FieldNum << "] to [" << FieldNum.getValue() << "] with count: " << C << " and data " << D << " bytes \n");
+          BBI->LastNode = NewNode;
+        }
+        else{
+          DEBUG_WITH_TYPE(DEBUG_TYPE_FRG, dbgs() << "No previous node found. It is the first node\n");
+          BBI->FirstNode = BBI->LastNode = NewNode;
+        }
+      }
+      else{
+        // Case that I is not struct access but a memory access
+        if (I.getOpcode() == Instruction::Load || I.getOpcode() == Instruction::Store){
+          DEBUG_WITH_TYPE(DEBUG_TYPE_FRG, dbgs() << "Found an instruction " << I << " is not a struct access but a load/store.\n");
+          if (BBI->LastNode == NULL){
+            // Create a dummy node for the first non-struct memory access
+            DEBUG_WITH_TYPE(DEBUG_TYPE_FRG, dbgs() << "Create a dummy node as the first node in the BB.\n");
+            BBI->FirstNode = BBI->LastNode = FRG->createNewNode();
+          }
+          else{
+            BBI->RemainBytes += getMemAccessDataSize(&I);
+            DEBUG_WITH_TYPE(DEBUG_TYPE_FRG, dbgs() << "Increment remaining data size to " << BBI->RemainBytes << "\n");
+          }
+        }
+      }
+    }
+    if (BBI->LastNode == NULL){
+      assert(BBI->FirstNode == NULL);
+      DEBUG_WITH_TYPE(DEBUG_TYPE_FRG, dbgs() << "Create a dummy node as BB does not have memory accesses\n");
+      BBI->FirstNode = BBI->LastNode = FRG->createNewNode();
+    }
+  }
+  // Before connect nodes, need to mark all back edges first
+  detectBackEdges(FRG, F);
+  // Connect nodes between different basic blocks
+  for (auto &BB : *F){
+    auto* BBI = FRG->getBasicBlockHelperInfo(&BB);
+    auto* Term = BB.getTerminator();
+    for (const auto *SB : Term->successors()){
+      auto* SBI = FRG->getBasicBlockHelperInfo(SB);
+      ExecutionCountType C = 0;
+      auto BBCount = StructInfo->getExecutionCount(&BB);
+      auto SBCount = StructInfo->getExecutionCount(SB);
+      if (BBCount.hasValue() && SBCount.hasValue()){
+        // Take probability of the branch
+        auto Prob = StructManager->getBranchProbability(&BB, SB);
+        assert(Prob.hasValue());
+        C = std::min(BBCount.getValue() * Prob.getValue(), SBCount.getValue() * 1.0);
+      }
+      if (!DisableIgnoreZeroCountNodes && C < 1e-3)
+        continue;
+      auto D = BBI->RemainBytes; // Use size of remaining data in BB
+      if (BBI->BackEdgeSet.find(SB) != BBI->BackEdgeSet.end())
+        FRG->connectNodes(BBI->LastNode, SBI->FirstNode, C, D, true);
+      else
+        FRG->connectNodes(BBI->LastNode, SBI->FirstNode, C, D);
```

Comment:
The above if-statement might be better as:

bool hasBackEdges = BBI->BackEdgeSet.find(SB) != BBI->BackEdgeSet.end();
FRG->connectNodes(BBI->LastNode, SBI->FirstNode, C, D, hasBackEdges);